File: C#/Leetcode/Math/SingleNumber.cs
namespace LeetcodeSolutions.Math
{
    // Leetcode 136
    // Submission Detail: https://leetcode.com/submissions/detail/142272930/
    // XOR
    public class SingleNumber
    {
        // Input: [-1,5,3,-1,5] -> 3
        // Input: [1]   -> 1
        // Tx = O(n)
        // Sx = O(1)
        public int FindSingleNumber(int[] nums)
        {
            int result = 0;

            // XOR. 0,1->1 && 1,0->1
            foreach (int num in nums)
                result ^= num;

            return result;
        }
    }
}

File: Java/Arrays/SingleNumber.java
package Arrays;

public class SingleNumber {
    // Leetcode 136: https://leetcode.com/problems/single-number/description/
    // Submission Detail: https://leetcode.com/problems/single-number/submissions/1680872330/

    //Input: nums = [2,2,1]
    //Output: 1

    //Input: nums = [4,1,2,1,2]
    //Output: 4

    //Input: nums = [1]
    //Output: 1

    // Intuition: xor operator when performed on 2 same numbers yields 0. And 0 xor on any number leads to the same number.
    public int singleNumber(int[] nums) {
        int nonDuplicate = 0;

        for(int num:nums) {
            nonDuplicate ^= num;
        }

        return nonDuplicate;
    }
}
